### ISSUE TITLE
RFC: re-opening the `Into<Option<_>>` debate :)

### ISSUE THREAD

**Issue by fengalin** [gtk-rs/gtk-rs-core#805]
## Another `Into<Option<_>>` experiment

In `gstreamer-rs` a [fix for nullability inconsistency] had an API changed induced an update to all the call sites for a function from`q.set_uri(uri)` to `q.set_uri(Some(uri))`. This revived the debate whether `Into<Option<_>>` should be used in argument position when the argument is optional.

This debate was [previously concluded]: "Into<Option<_>> considered harmful". The main arguments leading to this conclusion are:

- In some cases, this requires annotations, which can result messy, particularly   when dealing with closures.
- Users can be confused and not figure out that the API can also accept `None`.

We wanted to figure out whether the problems encountered at the time should still be considered harmefull and whether `Into<Option<_>>` should be reintroduce, at least in some cases, as we ended up doing for `set_uri`.

### TLDR

The TLDR is the above arguments are still valid:

- Type annotations are still necessary in some cases when we want to pass `None`. We probably don't want to use this with closures or complex function arguments.
- When users find code such as `q.set_uri(uri)`, they will probably not guess they can use `q.set_uri(None)` too.

Nevertheless, code looks nicer IMO when at least some of these arguments use this feature. I conducted an [experiment on `gstreamer-rs`] changing most candidates in manual code and updated `gst-plugins-rs` to use the resulting code base. See the difference in [this `gst-plugins-rs` commit].

In this issue, I'd like to show the result of a quick experimentation on `gtk-rs-core/glib` to illustrate the pros & cons. The changes and some illustrative test cases are available in [this `gtk-rs-core/glib` commit].

### The easy cases

There are easy cases for which there's no issue apart from the users not immediately figuring out that the arguments can be an `Option`.

#### `Copy + Clone` types

The `Copy + Clone` types handling is straightforward with limited impact on the function signature and body.

See the [`Channel::new`] implementation and the above test cases.

#### Reference to concrete types

When the argument's inner type is a reference to a concrete type, the implementation is quite straightforward too, with the addition of a lifetime.

See the [`DateTime::from_iso8601`] implementation and the above test cases.

#### Reference to a type by one of its trait implementation

This is where it starts to get tricky.

##### Strings

One very common case for this is when the argument is a string, like in `set_uri`. Currently, most functions use an `Option<&str>` for these kind of arguments. This allows using:

```rust
    q.my_function(Some("a `str` literal"));
    q.my_function(Some(&a_string)); // reference is mandatory
    q.my_function(None); // type for `None` is non-ambiguous
```

With the changes in [`ParamSpecBuilderExt::set_nick`], besides the same `Some` variants, we get:

```rust
    q.my_function("a `str` literal");
    q.my_function(&a_string); // reference is still mandatory
    q.my_function::<str>(None); // type for `None` needs disambiguation
```

The `None` case is unfortunate. It shows up here, but it actually was already an issue for others use cases which lead to the introduction of the `NONE` constant for some types.

The signature is a bit ugly, the `?Sized` bound is necessary to accept the `str` literal.

I tried to [have the signature accept plain `String`], but I gave up.

IMO the type annotation annoyance is acceptable compared to the usability improvement. There are many of these in [this `gst-plugins-rs` commit].

##### Subclasses

Another common use case involves subclasses. See [`SignalGroup::set_target`] as an example. In this case, with current API, we need to use the `NONE` constant:

```rust
    // type for `None` needs disambiguation
    SignalGroup::new(Object::static_type()).set_target(Object::NONE);
```

Disambiguation is still necessary with `Into<Option<_>>`, though we can use this instead:

```rust
    // type for `None` needs disambiguation
    SignalGroup::new(Object::static_type()).set_target::<Object>(None);
```

Which could render the `NONE` constants unnecessary. Note that this could be possible with current API if the type was declared as a generic argument `<T: IsA<_>>` instead of an `impl IsA<_>`.

IMO, when dealing with subclasses this is useful as it leads to leaner code. It might not be applicable to all functions though.

##### Functions and closures

The last example combines two types from their traits. One is a `Path`, so it is quite similar to the `str` case and the other is a function trait. This can be seen in the [`spawn_async`] implementation.

Of course, there's nothing special about the plain and `Some` cases. For the `None` case, we get to provide type annotation for the `None` argument:

```rust
    // type annotation needed for 1st arg due to `None`
    //                       v
    let _res = spawn_async::<path::Path, _>(
        None,
        &[path::Path::new("test")],
        &[],
        SpawnFlags::empty(),
        || {},
    );

    // type annotation needed for both args
    //                       v           v
    let _res = spawn_async::<path::Path, fn()>(
        None,
        &[path::Path::new("test")],
        &[],
        SpawnFlags::empty(),
        None,
    );
```

For some reason I can't immediately explain, using a `NONE` constant doesn't work:

```rust
    const SPAWN_ASYNC_FN_NONE: Option<fn()> = None;

    // Compilation fails, but shouldn't IMO:
    //                 cannot infer type v
    let _res = spawn_async::<path::Path, _>(
        None,
        &[path::Path::new("test")],
        &[],
        SpawnFlags::empty(),
        SPAWN_ASYNC_FN_NONE,
    );

```

This is getting tricky. The function signature is mimimalist here, but it would become unacceptable to impose a full signature to users who only want to pass `None` is the first place.

### Conclusion

IMO we should use `Into<Option<_>>` on a case by case basis:

- It was useful for the string arguments in all the cases I encoutered so far.
- It is also useful for the subclass traits args IMO. In the MR for the [experiment on `gstreamer-rs`], we started discussing cases where we may not want to do it.
- We should probably avoid this for the functions and closures args, except for really simple cases.

Apart from the manual code, applying this to automatically generated code could be useful, so a change to `gir` would be necessary.

[fix for nullability inconsistency]: https://gitlab.freedesktop.org/gstreamer/gstreamer-rs/-/commit/f9690817ad151ddcb5e038054a42988618cba7e5
[previously concluded]: https://github.com/gtk-rs/gir/issues/713
[experiment on `gstreamer-rs`]: https://gitlab.freedesktop.org/gstreamer/gstreamer-rs/-/merge_requests/1133
[this `gst-plugins-rs` commit]: https://gitlab.freedesktop.org/fengalin/gst-plugins-rs/-/commit/10f03c57c98c4a77af491b5b70936fb76f0b64fa
[this `gtk-rs-core/glib` commit]: https://github.com/fengalin/gtk-rs-core/commit/56651c084b2e7eb57b0336d3276a1653d3684a01
[`Channel::new`]: https://github.com/fengalin/gtk-rs-core/commit/56651c084b2e7eb57b0336d3276a1653d3684a01#diff-f1dce0d5b8c3a2313c9e4a292b5025c0a917a6dde00981654fbb7d24787c3613
[`DateTime::from_iso8601`]: https://github.com/fengalin/gtk-rs-core/commit/56651c084b2e7eb57b0336d3276a1653d3684a01#diff-80514931d7b3c8aeaca0c03880030e7895856daf12ed3912930e25fa8d89328f
[`ParamSpecBuilderExt::set_nick`]: https://github.com/fengalin/gtk-rs-core/commit/56651c084b2e7eb57b0336d3276a1653d3684a01#diff-1b97b457e5022fc4032d21e7382897ef99268ebe6b20ff3ee1e458b927ba84bc
[have the signature accept plain `String`]: https://gitlab.freedesktop.org/gstreamer/gstreamer-rs/-/merge_requests/1134#note_1600980
[`SignalGroup::set_target`]: https://github.com/fengalin/gtk-rs-core/commit/56651c084b2e7eb57b0336d3276a1653d3684a01#diff-c45c647601b3e764749f7907cc700d2e5503bbedb6f96e0e3f0888eff8643c9a
[`spawn_async`]: https://github.com/fengalin/gtk-rs-core/commit/56651c084b2e7eb57b0336d3276a1653d3684a01#diff-cf3e946adaf5f81ef984a00706c59e6fe900fbd6a7e789f92582977fc5af9789

**Comment by sdroege** [comment]
> The `None` case is unfortunate. It shows up here, but it actually was already an issue for others use cases which lead to the introduction of the `NONE` constant for some types.
> 
> The signature is a bit ugly, the `?Sized` bound is necessary to accept the `str` literal.

I think both of these parts are problematic. The signature is probably an obstacle for new users, as is the requirement to do something special for the `NONE` case. The `NONE` constants for objects that we currently have could help with the latter but not with the former.

Also for objects we have this problem anyway regardless of any of this because of the `IsA` trait :)

I'm wondering if for strings it wouldn't actually be better to use `impl Into<&str>`. Then you can use `None` as it is, and for string-ish types that are not `&str` you can use `&*val`, `val.as_ref()`, `val.as_str()` or similar. That is probably easier to discover and you have a more readable signature.

----

For functions and closures we currently always use boxed dyn closures when an `Option` is involved because otherwise the `None` is very hard to construct because any kind of closure and function pointer type is actually unique, and in case of closures unnameable.

Usually a better approach for improving these APIs is to provide a function with and without the optional closure argument.

Just some initial thoughts so far.

**Comment by sdroege** [comment]
The string discussion also has connections to what is planned in https://github.com/gtk-rs/gtk-rs-core/pull/600#issuecomment-1263209470 FWIW

**Comment by fengalin** [comment]
> The signature is probably an obstacle for new users, as is the requirement to do something special for the NONE case.

That's true, but with Rust expressive type system, "powerful" models can result difficult to understand to newcomers anyway. Luckily, we also have a powerful doc tooling (wink wink @GuillaumeGomez). With proper documentation, users should be able to figure out what can be done. Documentation shows up in the `rust-analyzer` popups, which I believe most users install. Some commonly used crates, such as `futures`, wouldn't be entirely understandable from the code itself IMO. I, for one, always open the documentation when I use a new crate. Only with poorly documented crates or when something does not behave as expected do I reach for the code.

**Comment by fengalin** [comment]
I gave this some more thoughts and made some changes to my experiment. This is also inspired by @sdroege's [introduction of `default()` for `Pipeline` & `Bin`](https://gitlab.freedesktop.org/gstreamer/gstreamer-rs/-/merge_requests/1136/diffs) as well as a [discussion we had](https://gitlab.freedesktop.org/gstreamer/gstreamer-rs/-/merge_requests/1134#note_1601051) regarding the possible introduction of an `unset_uri` to act like `set_uri(None)`.

The [resulting implementation](https://github.com/fengalin/gtk-rs-core/commit/e476cc203c11f6f385aad28795aac76351eda939) makes no more uses of any `Into<Option<_>>` and is based on builder-like patterns or additional functions, depending on the use case. I think this is more in-line with Rust philosophy.

### `Copy + Clone` types

The [`Channel` example](https://github.com/fengalin/gtk-rs-core/blob/e476cc203c11f6f385aad28795aac76351eda939/glib/src/main_context_channel.rs#L73) mimics the `Pipeline::default` approach, which seem obvious to me now: we are dealing with a constructor with one optional parameter.

### Reference to concrete types

For [`DateTime::from_iso8601`](https://github.com/fengalin/gtk-rs-core/blob/e476cc203c11f6f385aad28795aac76351eda939/glib/src/auto/date_time.rs#L25), the idea is similar: either we use the default (no time zone) or `DateTime::from_iso8601_with_tz`.

### Strings

It depends: either we are in either of the 2 above cases, or it's a case of `set_` / `unset_` or `discard_` or `reset_`.

By the way, in previous experiment, my use of the `PSpecBuilder` trait was not really significant as there was a `nick` setter that was designed to handle the `Some` case delagating to `set_nick` (the one I focussed on). In this particular example, I would advise to use an `unset_` or `discard_` function should the builder require the ability to unset something that was previously set before building.

### Subclasses

For the [`SignalGroup::set_target`](https://github.com/fengalin/gtk-rs-core/blob/e476cc203c11f6f385aad28795aac76351eda939/glib/src/gobject/auto/signal_group.rs#L22), this is again a case of `unset_` or `discard_` or `reset_` IMO.

### Functions and closures

The [`spawn_async`](https://github.com/fengalin/gtk-rs-core/blob/e476cc203c11f6f385aad28795aac76351eda939/glib/src/auto/functions.rs#L713) example was the one I though was too complicated for users in previous experiment. With the latest attempt, I used a pattern builder which I named `SpawnAsyncInvoker` since this is used to invoke a function, not to build an object. IMO, it is flexible enough and abstract the complexity from users.

This example applies to a free standing function, so I thought it was more straightforward to use this `Invoker` approach. For a method, we could just use a builder for the arguments to be passed to said method.

**Comment by jf2048** [comment]
First pass of some traits for string conversions here: https://github.com/gtk-rs/gtk-rs-core/pull/600/commits/89b8f3fec624d58dfe9a90d38abdb83f3a56a679 https://github.com/gtk-rs/gtk-rs-core/pull/600/commits/a267ae71180d2c21e982f21f961eb03d0d88dc5e

Ideally these should be preferred over `AsRef<str>`

**Comment by sdroege** [comment]
As we have `IntoOptionalGStr` and such things now, it might make sense to reconsider this. You'd have to pass a `None::<&str>` or similar anyway, so we could do the same for other types.

However we should also impl `IntoOptionalGStr` for `&GStr` / `&str` / etc.

@bilelmoussaoui Any opinions here?

**Comment by sdroege** [comment]
See https://github.com/gtk-rs/gtk-rs-core/pull/891

**Comment by GuillaumeGomez** [comment]
As mentioned previously, it will increase the gtk-rs code complexity and also make documentation harder to read. Is the gain really worth it?

**Comment by fengalin** [comment]
@GuillaumeGomez did you reply to the initial comment or to the refined: https://github.com/gtk-rs/gtk-rs-core/issues/805#issuecomment-1288079608

@slomo I'll try to take a look in the coming days. As I said in the comment I refered to above, I believe explicit methods to discard, unset, etc. would be valuable IMO.

**Comment by GuillaumeGomez** [comment]
I was answering to @sdroege based on your original arguments and small tests I ran. Based on your refined comment, there is no more use of `Into<Option<_>>`. If so, then my comment can be ignored.
